third_party/libpg_query/grammar/statements/select.y

```diff
@@ -2343,9 +2343,17 @@ c_expr:		columnref								{ $$ = $1; }
 				PGFuncCall *n = makeFuncCall(SystemFuncName("row"), $1, @1);
 				$$ = (PGNode *) n;
 			}
-			| '{' dict_arguments_opt_comma '}' {
-				PGFuncCall *n = makeFuncCall(SystemFuncName("struct_pack"), $2, @2);
-				$$ = (PGNode *) n;
+			| '{' dict_arguments_opt_comma '}' opt_indirection {
+				PGFuncCall *f = makeFuncCall(SystemFuncName("struct_pack"), $2, @2);
+                if ($4)
```

Comment:
This seems to be a pattern that is repeated in several places here - could we perhaps extract it into a separate node (`indirect_expr`), and do something like this:

```yacc

c_expr:
    ...
    | indirect_expr opt_indirection
    {
        if ($2) {
            PGAIndirection *n = makeNode(PGAIndirection);
            n->arg = (PGNode *) $1;
            n->indirection = check_indirection($2, yyscanner);
            $$ = (PGNode *) n;
        }
        else {
            $$ = (PGNode *) $1;
        }
    }
| ...

indirect_expr: 
    '?'
    {
        $$ = makeParamRef(0, @1);
    }
    | PARAM
    {
        PGParamRef *p = makeNode(PGParamRef);
        p->number = $1;
        p->location = @1;
         $$ = (PGNode *) n;
    }
    | ...    
```